core/editor/src/document/document_file.rs

```diff
@@ -1,143 +1,498 @@
-use crate::{consts::ROTATE_SNAP_INTERVAL, frontend::layer_panel::*, EditorError};
-use document_core::{document::Document as InternalDocument, layers::Layer, LayerId};
-use glam::{DAffine2, DVec2};
+pub use super::layer_panel::*;
+use crate::{frontend::layer_panel::*, EditorError};
+use document_core::{document::Document as InternalDocument, LayerId};
+use glam::{DAffine2, DVec2, Vec2Swizzles};
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 
+use crate::input::InputPreprocessor;
+use crate::message_prelude::*;
+use document_core::layers::BlendMode;
+use document_core::{DocumentResponse, Operation as DocumentOperation};
+use log::warn;
+
+use std::collections::VecDeque;
+
+use super::movement_handler::{MovementMessage, MovementMessageHandler};
+
+#[derive(PartialEq, Clone, Debug, Serialize, Deserialize, Hash)]
+pub enum FlipAxis {
+	X,
+	Y,
+}
+
+#[derive(PartialEq, Clone, Debug, Serialize, Deserialize, Hash)]
+pub enum AlignAxis {
+	X,
+	Y,
+}
+
+#[derive(PartialEq, Clone, Debug, Serialize, Deserialize, Hash)]
+pub enum AlignAggregate {
+	Min,
+	Max,
+	Center,
+	Average,
+}
+
 #[derive(Clone, Debug)]
-pub struct Document {
+pub struct DocumentMessageHandler {
 	pub document: InternalDocument,
+	pub document_backup: Option<InternalDocument>,
 	pub name: String,
 	pub layer_data: HashMap<Vec<LayerId>, LayerData>,
+	movement_handler: MovementMessageHandler,
 }
 
-impl Default for Document {
+impl Default for DocumentMessageHandler {
 	fn default() -> Self {
 		Self {
 			document: InternalDocument::default(),
+			document_backup: None,
 			name: String::from("Untitled Document"),
 			layer_data: vec![(vec![], LayerData::new(true))].into_iter().collect(),
+			movement_handler: MovementMessageHandler::default(),
 		}
 	}
 }
 
-impl Document {
+#[impl_message(Message, DocumentsMessage, Document)]
+#[derive(PartialEq, Clone, Debug)]
+pub enum DocumentMessage {
+	#[child]
+	Movement(MovementMessage),
+	DispatchOperation(Box<DocumentOperation>),
+	SelectLayers(Vec<Vec<LayerId>>),
+	SelectAllLayers,
+	DeselectAllLayers,
+	DeleteLayer(Vec<LayerId>),
+	DeleteSelectedLayers,
+	DuplicateSelectedLayers,
+	SetBlendModeForSelectedLayers(BlendMode),
+	SetOpacityForSelectedLayers(f64),
+	AddFolder(Vec<LayerId>),
+	RenameLayer(Vec<LayerId>, String),
+	ToggleLayerVisibility(Vec<LayerId>),
+	FlipSelectedLayers(FlipAxis),
+	ToggleLayerExpansion(Vec<LayerId>),
+	StartTransaction,
+	RollbackTransaction,
+	AbortTransaction,
+	CommitTransaction,
+	ExportDocument,
+	RenderDocument,
+	Undo,
+	NudgeSelectedLayers(f64, f64),
+	FlipLayer(Vec<LayerId>, bool, bool),
```

Comment:
I believe this message is removed in master, replaced by `FlipSelectedLayers`. Also, `FlipSelectedLayers` should probably be down here, next to the other transformations of the selected layers.